apps/intellij-plugin/src/main/kotlin/ai/hash/rustextra/projectView/ModuleGroupTreeStructureProvider.kt

```diff
@@ -0,0 +1,77 @@
+package ai.hash.rustextra.projectView
+
+import com.intellij.ide.projectView.TreeStructureProvider
+import com.intellij.ide.projectView.ViewSettings
+import com.intellij.ide.util.treeView.AbstractTreeNode
+import com.intellij.openapi.project.DumbAware
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiFile
+
+class ModuleGroupTreeStructureProvider : TreeStructureProvider, DumbAware {
+    override fun modify(
+        parent: AbstractTreeNode<*>,
+        children: MutableCollection<AbstractTreeNode<*>>,
+        settings: ViewSettings?,
+    ): MutableCollection<AbstractTreeNode<*>> {
+        val nodes = ArrayList<AbstractTreeNode<*>>()
+
+        val directories = HashMap<String, AbstractTreeNode<PsiDirectory>>()
+        val files = HashMap<String, AbstractTreeNode<PsiFile>>()
+
+        // To merge all `module` with their corresponding `module.rs` file we need to do two iterations
+        // ... first collect all directories and files via lookup tables
+        for (child in children) {
+            val value = child.value
+
+            if (value is PsiDirectory) {
+                // Statement above checks that this is the case
+                @Suppress("UNCHECKED_CAST")
+                directories[value.name] = child as AbstractTreeNode<PsiDirectory>
+            }
+
+            // all rust files need to end with `.rs`, therefore it is safe to just look for `*.rs` files
+            // this way we do not need to require an implicit dependency on the `Rust` plugin to provide us
+            // with the correct `value.virtualFile.fileType` value
```

Comment:
Would depending on Rust have other benefits?

Reply:
For now, no, I was initially dependent on Rust, but plugins (and Java, for that fact) are unlike Rust dependencies. You need to pin a specific version. This makes it super complicated because the Rust plugin actually publishes **2-3** distinct versions per release. (You can see it here: https://plugins.jetbrains.com/plugin/8182-rust/versions/stable). This would also mean that our plugin would need to update every time their plugin updates, which is a headache (otherwise our plugin would get disabled). We can indirectly influence the plugin using JetBrain plugin internals, but should the occasion arise for a specific feature that needs a direct dependency I am very open to it! :D

Reply:
Thanks for the clarification!
Also, I wonder if we want to add the same behavior to TS files at some point. Depending on the Rust plugin would be suboptimal I think.